src/analysis/syntax_doc.ml

```diff
@@ -0,0 +1,229 @@
+        match typ_private with
+        | Public -> (r_name, r_description, r_url)
+        | Private ->
+            ( Format.sprintf "Private %s" r_name,
+              Format.sprintf
+                "%s This type is private, values cannot be constructed \
+                 directly but can be de-structured as usual."
+                r_description,
+              "privatetypes.html#ss:private-types-variant" )
+      in
+      Some { name; description; documentation = syntax_doc_url url }
+  | (_, Type_kind (Ttype_variant _))
+    :: (_, Type_declaration { typ_private = Public; _ })
+    :: _ ->
+      Some
+        {
+          name = "Empty Variant Type";
+          description = "An empty variant type.";
+          documentation = syntax_doc_url "emptyvariants.html";
+        }
+  | (_, Type_kind Ttype_abstract)
+    :: (_, Type_declaration { typ_private = Public; typ_manifest = None; _ })
+    :: _ ->
+      Some
+        {
+          name = "Abstract Type";
+          description =
+            "Define variants with arbitrary data structures, including other \
+             variants, records, and functions";
+          documentation = syntax_doc_url "typedecl.html#ss:typedefs";
+        }
+  | (_, Type_kind Ttype_abstract)
+    :: (_, Type_declaration { typ_private = Private; _ })
+    :: _ ->
+      Some
+        {
+          name = "Private Type Abbreviation";
+          description =
+            "Declares a type that is distinct from its implementation type \
+             `typexpr`.";
+          documentation =
+            syntax_doc_url "privatetypes.html#ss:private-types-abbrev";
+        }
+  | (_, Expression _)
+    :: (_, Expression _)
+    :: (_, Value_binding _)
+    :: (_, Structure_item ({ str_desc = Tstr_value (Recursive, _); _ }, _))
+    :: _ ->
+      Some
+        {
+          name = "Recursive value definition";
+          description =
+            "Supports a certain class of recursive definitions of \
+             non-functional values.";
+          documentation = syntax_doc_url "letrecvalues.html";
+        }
+  | (_, Module_expr _) :: (_, Module_type { mty_desc = Tmty_typeof _; _ }) :: _
+    ->
+      Some
+        {
+          name = "Recovering module type";
+          description =
+            "Expands to the module type (signature or functor type) inferred \
+             for the module expression `module-expr`. ";
+          documentation = syntax_doc_url "moduletypeof.html";
+        }
+  | (_, Module_expr _)
+    :: (_, Module_expr _)
+    :: (_, Module_binding _)
+    :: (_, Structure_item ({ str_desc = Tstr_recmodule _; _ }, _))
+    :: _ ->
+      Some
+        {
+          name = "Recursive module";
+          description =
+            "A simultaneous definition of modules that can refer recursively \
+             to each others.";
+          documentation = syntax_doc_url "recursivemodules.html";
+        }
+  | (_, Expression _)
+    :: (_, Expression _)
+    :: (_, Expression _)
```

Comment:
This has a `Case` node upstream. Why the change?

Reply:
There was [another commit](https://github.com/ocaml/merlin/commit/fb180ed04dbdd1cdc9157f34af905a75c748ecbd) on the upstream to remove the `Case` node when porting to OCaml 5. I believe the reason for this change is n-ary functions, but I'm not 100% confident for that being the reason. Either way, this change was needed to make the tests pass.